\documentclass{beamer}

\usepackage[latin1]{inputenc}
\usepackage{hyperref}
\usepackage{multirow}
\usetheme{Copenhagen}
\title[PaGE in Python]{PaGE in Python}

\institute{University of Pennsylvania}
\date{October 19, 2012}

\usepackage{listings}

\begin{document}

\begin{frame}
  \titlepage
\end{frame}

\begin{frame}{Agenda}
  \tableofcontents
\end{frame}

\section{What is PaGE?}

\begin{frame}
  \begin{centering}
  {\Large{What is PaGE?}}    
  \end{centering}
\end{frame}

\begin{frame}{What is PaGE?}
  \begin{itemize}
  \item {\bf Pa}tterns from {\bf G}ene {\bf E}xpression
  \item Discover sets of differentially expressed genes from
    microarray intensity data
  \item Written by Gregory Grant
  \item ~5,000 line Perl program
  \end{itemize}
\end{frame}

\begin{frame}{Input file}
  
  PaGE's input is a big 2-D array:
  \begin{itemize}
  \item Cells are intensities of gene expression
  \item Each row is a 'feature' (gene)
  \item Each column is a replicate
  \item Columns are grouped into conditions
  \end{itemize}
  \begin{tabular}{|l|lll|l|lll|}
    & \multicolumn{3}{|c|}{Condition 1} & ... & \multicolumn{3}{|c|}{Condition $N$}\\
    & Rep. 1 & ... & Rep. $n_1$       & ... & Rep 1 & ... & Rep. $n_{N}$\\
    \hline\\
    Gene 1 & 0.39 & ... & 4.53       & ... & 3.49 & ... & 1.06 \\
    Gene 2 & 1.33 & ... & 2.40       & ... & 0.78 & ... & 1.85 \\
    ... & ... & ... & ...       & ... & ... & ... & ... \\
    Gene M & 1.33 & ... & 2.40       & ... & 0.78 & ... & 1.85 \\
    
  \end{tabular}
  
\end{frame}

\begin{frame}{Why mess with it?}
  \begin{itemize}
  \item Originally written for microarray data
  \item In order to use for RNA-Seq, need to normalize measurements
    (since we may have a different sample size for each replicate)
  \item Current implementation doesn't handle normalization in an ideal way
  \item Improve performance for very large input files
  \end{itemize}
\end{frame}

\section{Highlights of new version}
\begin{frame}
  \begin{centering}
    {\Large{Highlights of new version}}
  \end{centering}
\end{frame}

\begin{frame}{Faster numerical operations}
  We should be able to speed it up using {\em NumPy}:
  \begin{itemize}
  \item Stands for {\bf Num}erical {\bf Py}thon something or other...
  \item Efficient storage and access to large multidimensional arrays
  \item Provides very fast numerical operations
  \item Pushes work down into in highly optimized C
  \item Lots of numerical and scientific functions
  \item Integrates well with other Python libraries
  \end{itemize}
\end{frame}

\begin{frame}{Other highlights}
  While we're at it, we can take advantage of some other libraries.
\end{frame}

\begin{frame}{Command-line parsing}
    \begin{itemize}
    \item Python's {\em argparse} library is great.
    \item Handles basic parsing of command line
    \item Converts strings to ints, floats, files, bools, etc...
    \item Does some validation
    \item Automatically generates usage information and help/error messages
    \end{itemize}
\end{frame}

\begin{frame}{Tests and documentation}
  \begin{itemize}
  \item Test one function at a time
  \item Writing tests first, based on output of Perl version
  \item Then writing Python code to match output
  \item Document functions using docstrings
  \item Some tests embedded in documentation
  \end{itemize}  
\end{frame}

\section{Demo}
\begin{frame}
  \begin{centering}
    \Large{Demo}        
  \end{centering}
\end{frame}

\begin{frame}{Demo}
  \begin{itemize}
  \item Run original PaGE, look at output
  \item Tour of new PaGE directory structure

  \item Run new PaGE
    \begin{itemize}
    \item Show usage info
    \item Show full help message
    \item Show with bad in file
    \item Show with bad num channels
    \end{itemize}

  \item Show command-line parsing code

  \item Walk through tstat in iPython
    \begin{itemize}
    \item Show vectorization
    \end{itemize}

  \item Compare timings

  \item Show pydoc

  \item Show test for all\_subsets
  \item Show test for tstat
  \end{itemize}
\end{frame}

\end{document}
